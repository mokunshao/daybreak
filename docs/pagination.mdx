---
title: Pagination
---

import {useState} from 'react';
import { Playground, Props } from 'docz';
import Pagination from '../src/pagination/pagination.tsx';

# Pagination

## Basic usage

<Playground>
{() => {
  const [current, setCurrent] = useState(1);
  const [current2, setCurrent2] = useState(1);
  const onChange = (n) => setCurrent(n);
  const onChange2 = (n) => setCurrent2(n);
  return (
    <div>
      <Pagination pageSize={10} total={200} current={current} onChange={onChange} />
      <br />
      <br />
      <Pagination pageSize={10} total={40} current={current2} onChange={onChange2} />
    </div>
  );
}}
</Playground>

## Props

<Props of={Pagination} />
