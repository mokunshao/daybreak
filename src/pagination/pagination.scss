.daybreak-pagination {
  display: inline-flex;
  flex-direction: row;

  &-button {
    margin-left: 0.1em;
    margin-right: 0.1em;
    display: inline-flex;
    align-items: center;
  }
}
