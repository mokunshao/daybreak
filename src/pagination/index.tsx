import { Pagination } from './pagination';

export default Pagination;

export { Pagination };
