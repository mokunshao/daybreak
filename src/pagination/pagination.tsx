import React, {
  HTMLProps, useMemo, MouseEventHandler, useState,
} from 'react';
import { joinedClass } from '../utils/joinedClass';
import { classes } from '../utils/classes';
import './pagination.scss';
import { Button } from '../button/button';
import Icon from '../icon/icon';

const baseClass = joinedClass('pagination');

interface Props extends Omit<HTMLProps<HTMLDivElement>, 'onChange'> {
  current: number;
  pageSize: number;
  total: number;
  onChange: Function;
  disabled?: boolean;
}

const Pagination: React.FC<Props> = (props) => {
  const {
    className, total, pageSize, current, onChange,
  } = props;
  const pagesCount = useMemo(() => total / pageSize, [total, pageSize]);
  const arr = [];
  for (let index = 0; index < pagesCount; index += 1) {
    arr.push(index);
  }

  const handleClick = (n: number) => {
    onChange(n);
  };
  return (
    <div className={classes(baseClass(), className)}>
      <Button
        className={baseClass('button')}
        type="button"
        onClick={() => handleClick(current - 1)}
      >
        <Icon name="left" />
      </Button>
      {arr.map((item) => (
        <Button
          className={baseClass('button')}
          type="button"
          key={item}
          mode={current === item ? 'primary' : 'normal'}
          onClick={() => handleClick(item)}
        >
          {item}
        </Button>
      ))}
      <Button className={baseClass('button')} type="button">
        <Icon
          name="right"
          onClick={() => handleClick(current + 1)}
        />
      </Button>
    </div>
  );
};

export default Pagination;

export { Pagination };
