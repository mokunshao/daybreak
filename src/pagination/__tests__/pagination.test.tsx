import React, { useState } from 'react';
import { render } from '@testing-library/react';
import { Pagination } from '../pagination';

function Example() {
  const [current, setCurrent] = useState(1);
  const onChange = (n: number) => setCurrent(n);
  return (
    <div>
      <Pagination pageSize={10} total={200} current={current} onChange={onChange} />
    </div>
  );
}


describe('Pagination', () => {
  it('exits', () => {
    expect(<Example />).toBeTruthy();
  });

  it('match snapshot', () => {
    const { asFragment } = render(<Example />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('click', () => {
    const { getByText } = render(<Example />);
    const button = getByText('2');
    expect(button.classList.contains('daybreak-button-primary')).toBeFalsy();
    button.click();
    expect(button.classList.contains('daybreak-button-primary')).toBeTruthy();
    const button2 = getByText('4');
    button2.click();
    expect(button.classList.contains('daybreak-button-primary')).toBeFalsy();
    expect(button2.classList.contains('daybreak-button-primary')).toBeTruthy();
  });
});
