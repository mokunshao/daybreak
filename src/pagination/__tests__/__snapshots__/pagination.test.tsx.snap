// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="daybreak-pagination"
    >
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        disabled=""
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          <svg
            class="daybreak-icon"
          >
            <use
              xlink:href="#left"
            />
          </svg>
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-primary daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          1
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          2
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          3
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          4
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          5
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          ...
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          20
        </div>
      </button>
      <button
        class="daybreak-button daybreak-button-normal daybreak-pagination-button"
        type="button"
      >
        <div
          style="opacity: 1;"
        >
          <svg
            class="daybreak-icon"
          >
            <use
              xlink:href="#right"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
